R/get_stats.R

```diff
@@ -133,39 +133,58 @@ getRhythmStats = function(
 #'
 #' @export
-getDiffRhythmStats = function(fit, rhyStats, condLevels) {
-  cond = .SD = peak_phase = trough_phase = NULL
+getDiffRhythmStats = function(fit, rhyStats, condLevels = fit$condLevels) {
+  cond = .SD = diff_peak_phase = diff_trough_phase = condInt = . = cond1  =
+    condInt1 = mean_value1 = peak_trough_amp1 = rms_amp1 = peak_phase1 =
+    trough_phase1 = cond2 = condInt2 = mean_value2 = peak_trough_amp2 =
+    rms_amp2 = peak_phase2 = trough_phase2 = feature = NULL
 
   assertClass(fit, 'limorhyde2')
   assertTRUE(fit$nConds >= 2)
   assertDataTable(rhyStats)
   assertTRUE(attr(rhyStats, 'statType') == 'rhy')
   assertTRUE('cond' %in% colnames(rhyStats))
-  assertAtomicVector(condLevels, len = 2L)
   assertSubset(condLevels, fit$condLevels)
   assertSubset(condLevels, levels(rhyStats$cond))
 
   d0 = rhyStats[cond %in% condLevels]
   set(d0, j = 'cond', value = factor(d0$cond, condLevels))
-  data.table::setorderv(d0, 'cond')
+  d0[, condInt := as.integer(cond)]
 
   fitType = attr(rhyStats, 'fitType')
   byCols = c('feature', if (fitType == 'posterior_samples') 'posterior_sample')
-  cols = c('mean_value', 'peak_trough_amp', 'rms_amp', 'peak_phase', 'trough_phase')
 
-  diffRhyStats = d0[, lapply(.SD, diff), by = byCols, .SDcols = cols]
-  diffRhyStats[, peak_phase := centerCircDiff(peak_phase, fit$period)]
-  diffRhyStats[, trough_phase := centerCircDiff(trough_phase, fit$period)]
-  data.table::setnames(diffRhyStats, cols, paste0('diff_', cols))
+  diffRhyStats = merge(
+    d0, d0, allow.cartesian = TRUE, by = byCols, suffixes = 1:2)
+  diffRhyStats = diffRhyStats[condInt1 < condInt2]
+  diffRhyStats = diffRhyStats[,
+    .(cond1, cond2,
+      diff_mean_value = mean_value2 - mean_value1,
+      diff_peak_trough_amp = peak_trough_amp2 - peak_trough_amp1,
+      diff_rms_amp = rms_amp2 - rms_amp1,
+      diff_peak_phase = peak_phase2 - peak_phase1,
+      diff_trough_phase = trough_phase2 - trough_phase1),
+    by = byCols]
+
+  diffRhyStats[, diff_peak_phase := centerCircDiff(diff_peak_phase, fit$period)]
+  diffRhyStats[, diff_trough_phase := centerCircDiff(diff_trough_phase, fit$period)]
+
+  #create data.table of conditions
+  pairDt = unique(diffRhyStats[, .(cond1, cond2)])
 
   # calculate rms difference in rhythmic fit between conditions
   featureIdx = rownames(fit$coefficients) %in% unique(rhyStats$feature)
-  rmsDiffRhy = getRmsDiffRhy(fit, condLevels, fitType, featureIdx)
-  diffRhyStats = merge(diffRhyStats, rmsDiffRhy, sort = FALSE)
+  rmsDiffRhy = foreach(cond1 = pairDt$cond1, cond2 = pairDt$cond2,
+                       .combine = rbind) %do% {
+    rmsDiffRhyTmp = getRmsDiffRhy(fit, c(cond1, cond2), fitType, featureIdx)
+    rmsDiffRhyTmp = cbind(rmsDiffRhyTmp, cond1, cond2)
+    rmsDiffRhyTmp}
+  diffRhyStats = merge(
+    diffRhyStats, rmsDiffRhy, sort = FALSE, by = c(byCols, 'cond1', 'cond2'))
+  data.table::setorderv(diffRhyStats, c(byCols, 'cond1', 'cond2'))
```

Comment:
Originally, the code was intentionally not sorting, so the features would stay in the same order in which they were provided. I think I'd like to keep it that way for now, so let's trying removing this line. As is, if we wanted to keep this line, we could achieve the same thing by just removing `sort = FALSE` in the `merge` command.